src/Data/Nat/Binary/Minus.agda

```diff
@@ -0,0 +1,239 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Subtraction on Bin and some of its property proofs.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.Nat.Binary.Minus where
+
+open import Data.Nat.Binary.Base
+open import Data.Nat.Binary.Properties
+open import Data.Empty using (⊥)
+open import Data.Nat as ℕ using (ℕ)
+import Data.Nat.Properties as ℕₚ
+open import Data.Product using (_×_; _,_; proj₁; proj₂; ∃)
+open import Data.Sum using (inj₁; inj₂)
+open import Function using (_∘_; _$_)
+open import Relation.Binary using (Tri; _Preserves_⟶_; _Preserves₂_⟶_⟶_)
+open import Relation.Binary.PropositionalEquality as P using
+  (_≡_; _≢_; _≗_; refl; sym; trans; cong; cong₂; resp₂; subst)
+import Relation.Binary.Reasoning.Base.Triple as InequalityReasoning
+open import Relation.Nullary using (Dec; yes; no)
+open import Relation.Nullary.Negation using (contradiction)
+open import Algebra.FunctionProperties {A = ℕᵇ} _≡_
+open import Algebra.Properties.CommutativeSemigroup +-commSemigroup
+     using (xy∙z≈y∙xz; x∙yz≈y∙xz; xy∙zu≈xz∙yu)
+
+private
+  module IneqReasoning = InequalityReasoning {A = ℕᵇ} {_≈_ = _≡_} ≤-isPreorder
+                                   <-trans (resp₂ _<_) <⇒≤ <-≤-trans ≤-<-trans
+open ℕᵇ
+open Tri
+open _<_
+
+Minus : ℕᵇ → ℕᵇ → Set                                 -- result of subtraction
+Minus x y =  ℕᵇ × Tri (x < y) (x ≡ y) (x > y)
+
+minus :  ∀ x y → Minus x y                            -- extended subtraction
+minus zero     zero       =  (0ᵇ ,       tri-≈ refl)
+minus zero     (2[1+ _ ]) =  (0ᵇ ,       tri-< 0<even)
+minus zero     1+[2 _ ]   =  (0ᵇ ,       tri-< 0<odd)
+minus 2[1+ x ] zero       =  (2[1+ x ] , tri-> 0<even)
+minus 2[1+ x ] 2[1+ y ]   with minus x y
+... | (d , cmp) = (double d , even<even-mono-cmp cmp)
+minus 2[1+ x ] 1+[2 y ]  with minus x y
+... | (_ , tri< x<y _   _)   =  (0ᵇ  , tri-< (even<odd x<y))
+... | (d , tri≈ _   x≡y _)   =  (1+[2 d ] , tri-> (odd<even (inj₂ (sym x≡y))))
+... | (d , tri> _   _   x>y) =  (1+[2 d ] , tri-> (odd<even (inj₁ x>y)))
+minus 1+[2 x ] zero      =  (1+[2 x ] , tri-> 0<odd)
+minus 1+[2 x ] 2[1+ y ]  with minus x y
+... | (_ , tri< x<y _  _)   =  (0ᵇ , tri-< (odd<even (inj₁ x<y)))
+... | (_ , tri≈ _   eq _)   =  (0ᵇ , tri-< (odd<even (inj₂ eq)))
+... | (d , tri> _   _  x>y) =  (pred (double d) , tri-> (even<odd x>y))
+minus 1+[2 x ] 1+[2 y ]  with minus x y
+... | (d , cmp) =  (double d , odd<odd-mono-cmp cmp)
+
+infixl 6 _∸_
+
+_∸_ : Op₂ ℕᵇ              -- traditional subtraction
+_∸_ x =  proj₁ ∘ minus x
+
+0∸x≡0 :  ∀ x → 0ᵇ ∸ x ≡ 0ᵇ
+0∸x≡0 zero     =  refl
+0∸x≡0 2[1+ _ ] =  refl
+0∸x≡0 1+[2 _ ] =  refl
```

Comment:
For example, with the new definition I propose above no pattern matching is required here. It simply becomes `0∸x≡0 x =  refl`

Reply:
Done. I remove this lemma, instead one has to apply ``refl``.

Reply:
Hmm the lemma still seems useful to have as one may not realise that `refl` is the answer. I'd probably keep it in.